apollo-integration/src/test/java/com/apollographql/apollo/ApolloInterceptorChainTest.java

```diff
@@ -0,0 +1,357 @@
+
+import javax.annotation.Nonnull;
+
+import okhttp3.MediaType;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.ResponseBody;
+
+import static com.apollographql.apollo.interceptor.ApolloInterceptor.CallBack;
+import static com.apollographql.apollo.interceptor.ApolloInterceptor.InterceptorResponse;
+import static com.google.common.truth.Truth.assertThat;
+
+public class ApolloInterceptorChainTest {
+
+  public static final int TIMEOUT_SECONDS = 2;
+  private ApolloInterceptorChain chain;
+
+
+  @After
+  public void tearDown() {
+    chain = null;
+  }
+
+  @Test
+  public void onProceedCalled_chainPassesControlToInterceptor() throws ApolloException, TimeoutException, InterruptedException {
+
+    final Counter counter = new Counter(1);
+
+    EpisodeHeroName query = createQuery();
+
+    ApolloInterceptor interceptor = new ApolloInterceptor() {
+      @Nonnull @Override
+      public InterceptorResponse intercept(Operation operation, ApolloInterceptorChain chain) throws ApolloException {
+        counter.decrement();
+        return null;
+      }
+
+      @Override
+      public void interceptAsync(@Nonnull Operation operation, @Nonnull ApolloInterceptorChain chain, @Nonnull ExecutorService dispatcher, @Nonnull CallBack callBack) {
+
+      }
+
+      @Override public void dispose() {
+
+      }
+    };
+
+    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
+    chain = new RealApolloInterceptorChain(query, interceptors);
+
+    chain.proceed();
+
+    if (!counter.isZero()) {
+      Assert.fail("Control not passed to the interceptor");
+    }
+  }
+
+  @Test
+  public void onProceedAsyncCalled_chainPassesControlToInterceptor() throws TimeoutException, InterruptedException {
+
+    final NamedCountDownLatch latch = new NamedCountDownLatch("latch", 1);
+    EpisodeHeroName query = createQuery();
+
+    ApolloInterceptor interceptor = new ApolloInterceptor() {
+      @Nonnull @Override
+      public InterceptorResponse intercept(Operation operation, ApolloInterceptorChain chain) throws ApolloException {
+        return null;
+      }
+
+      @Override
+      public void interceptAsync(@Nonnull Operation operation, @Nonnull ApolloInterceptorChain chain, @Nonnull ExecutorService dispatcher, @Nonnull CallBack callBack) {
+        latch.countDown();
+      }
+
+      @Override public void dispose() {
+
+      }
+    };
+
+    List<ApolloInterceptor> interceptors = Collections.singletonList(interceptor);
+    chain = new RealApolloInterceptorChain(query, interceptors);
+
+    chain.proceedAsync(Executors.newFixedThreadPool(1), new CallBack() {
+      @Override public void onResponse(@Nonnull InterceptorResponse response) {
+
+      }
+
+      @Override public void onFailure(@Nonnull ApolloException e) {
+
+      }
+    });
+
+    //Latch's count should go down to zero, else timeout is reached
+    //which means the test fails.
+    latch.awaitOrThrowWithTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS);
```

Comment:
Replaced latch with a simple AtomicInteger based counter, just to ensure that the callbacks are indeed being called, in spite of using `Utils.immediateExecutorService`.